src/sys/windows/afd.rs

```diff
@@ -0,0 +1,184 @@
+        f.debug_struct("AfdPollInfo").finish()
+    }
+}
+
+impl Afd {
+    pub fn new(cp: &CompletionPort) -> io::Result<Afd> {
+        let mut afd_helper_name = OsStr::new(AFD_HELPER_NAME)
+            .encode_wide()
+            .collect::<Vec<_>>();
+
+        let mut afd_helper_handle: HANDLE = INVALID_HANDLE_VALUE;
+        let mut iosb: IO_STATUS_BLOCK = unsafe { std::mem::zeroed() };
+
+        unsafe {
+            let mut objname = UNICODE_STRING {
+                // Lengths are calced in bytes
+                Length: (afd_helper_name.len() * 2) as u16,
+                MaximumLength: (afd_helper_name.len() * 2) as u16,
+                Buffer: afd_helper_name.as_mut_ptr(),
+            };
+            let mut afd_helper_attributes = OBJECT_ATTRIBUTES {
+                Length: size_of::<OBJECT_ATTRIBUTES>() as ULONG,
+                RootDirectory: null_mut() as HANDLE,
+                ObjectName: &mut objname,
+                Attributes: 0 as ULONG,
+                SecurityDescriptor: null_mut() as PVOID,
+                SecurityQualityOfService: null_mut() as PVOID,
+            };
+            let status = NtCreateFile(
+                &mut afd_helper_handle,
+                SYNCHRONIZE,
+                &mut afd_helper_attributes,
+                &mut iosb,
+                null_mut(),
+                0 as ULONG,
+                FILE_SHARE_READ | FILE_SHARE_WRITE,
+                FILE_OPEN,
+                0 as ULONG,
+                null_mut(),
+                0 as ULONG,
+            );
+            if status != STATUS_SUCCESS {
+                return Err(io::Error::from_raw_os_error(
+                    RtlNtStatusToDosError(status) as i32
+                ));
+            }
+            let fd = File::from_raw_handle(afd_helper_handle as RawHandle);
+            let afd = Afd { fd };
+            let token = NEXT_TOKEN.fetch_add(1, Ordering::Relaxed) + 1;
+            cp.add_handle(token, &afd.fd)?;
+            match SetFileCompletionNotificationModes(
+                afd_helper_handle,
+                FILE_SKIP_SET_EVENT_ON_HANDLE,
+            ) {
+                0 => Err(io::Error::last_os_error()),
+                _ => Ok(afd),
+            }
+        }
+    }
+
+    pub fn poll(
+        &self,
+        info: &mut AfdPollInfo,
+        iosb: &mut IO_STATUS_BLOCK,
+        apccontext: PVOID,
+    ) -> io::Result<bool> {
+        unsafe {
+            let info_ptr: PVOID = info as *mut _ as PVOID;
+            iosb.u.Status = STATUS_PENDING;
+            let status = NtDeviceIoControlFile(
+                self.fd.as_raw_handle(),
+                null_mut(),
+                None,
+                apccontext,
+                iosb,
+                IOCTL_AFD_POLL,
+                info_ptr,
+                size_of::<AfdPollInfo>() as u32,
+                info_ptr,
+                size_of::<AfdPollInfo>() as u32,
+            );
+            match status {
```

Comment:
In what cases here does `NtDeviceIoControlFile` **not** take ownership of `apccontext`? My guess is, at the very least, it does not take ownership when there is an error.

We need to take care to not leak memory here.

Reply:
It looks like it just copy paste the apccontext to iocp as a Overlapped pointer. Not using it at all.

Reply:
And I confirmed we aren't leaking memory using Dr.Memory.

Reply:
main.rs
```
use std::io;
use std::net::{SocketAddr, Ipv4Addr};
use std::thread::sleep;
use std::time::Duration;

use env_logger;
use mio::{Poll, Token, Interests};
use mio::net::UdpSocket;

fn main() -> io::Result<()> {
    env_logger::init();

    let target = SocketAddr::new(Ipv4Addr::new(192,168,1,1).into(), 53);
    let buf = &[0x00; 1400];
    let socket = UdpSocket::bind(SocketAddr::new(Ipv4Addr::UNSPECIFIED.into(), 0))?;

    let poll = Poll::new()?;
    poll.registry().register(&socket, Token(0), Interests::READABLE)?;

    drop(socket.send_to(buf, target));
    sleep(Duration::from_millis(1000));

    Ok(())
}
```
result.txt
```
Dr. Memory version 2.2.0 build 1 built on Jul  1 2019 00:40:18
Windows version: WinVer=105;Rel=1903;Build=18936;Edition=Professional
Dr. Memory results for pid 31512: "mio-mao.exe"
Application cmdline: ".\target\debug\mio-mao.exe"
Recorded 117 suppression(s) from default C:\Program Files (x86)\Dr. Memory\bin64\suppress-default.txt

Error #1: UNINITIALIZED READ: reading register eax
# 0 WS2_32.dll!WSALookupServiceNextW                                          +0x3031   (0x00007ffdca6eedd1 <WS2_32.dll+0xedd1>)
# 1 WS2_32.dll!WSALookupServiceNextW                                          +0x3c3d   (0x00007ffdca6ef9de <WS2_32.dll+0xf9de>)
# 2 WS2_32.dll!WSAStartup                                                     +0x33c    (0x00007ffdca6efd4d <WS2_32.dll+0xfd4d>)
# 3 std::sync::once::{{impl}}::call_once::{{closure}}<>                        [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\sync\once.rs:220]
# 4 std::sync::once::Once::call_inner                                          [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\sync\once.rs:387]
# 5 std::sys_common::net::UdpSocket::bind                                      [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\sys_common\net.rs:468]
# 6 core::ops::function::FnMut::call_mut<> core::result::Result<>,(core::result::Result<>)> [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libcore\ops\function.rs:148]
# 7 std::net::each_addr<> core::result::Result<>,std::sys_common::net::UdpSocket> [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\net\mod.rs:113]
# 8 std::net::udp::UdpSocket::bind<>                                           [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\net\udp.rs:94]
# 9 mio::net::udp::UdpSocket::bind                                             [C:\Users\dennis\.cargo\git\checkouts\mio-76192d34136afce1\7358ab7\src\net\udp.rs:122]
#10 mio_mao::main                                                              [G:\Dev\test-mio-mao\src\main.rs:15]
#11 std::rt::lang_start::{{closure}}<>                                         [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
#12 std::rt::lang_start_internal                                               [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\rt.rs:48]
#13 std::rt::lang_start<>                                                      [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
#14 main       
Note: @0:00:01.069 in thread 31520
Note: instruction: cmp    %eax %r12d

Error #2: UNINITIALIZED READ: reading register eflags
# 0 WS2_32.dll!WSALookupServiceNextW                                          +0x304f   (0x00007ffdca6eedef <WS2_32.dll+0xedef>)
# 1 WS2_32.dll!WSALookupServiceNextW                                          +0x3c3d   (0x00007ffdca6ef9de <WS2_32.dll+0xf9de>)
# 2 WS2_32.dll!WSAStartup                                                     +0x33c    (0x00007ffdca6efd4d <WS2_32.dll+0xfd4d>)
# 3 std::sync::once::{{impl}}::call_once::{{closure}}<>                        [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\sync\once.rs:220]
# 4 std::sync::once::Once::call_inner                                          [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\sync\once.rs:387]
# 5 std::sys_common::net::UdpSocket::bind                                      [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\sys_common\net.rs:468]
# 6 core::ops::function::FnMut::call_mut<> core::result::Result<>,(core::result::Result<>)> [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libcore\ops\function.rs:148]
# 7 std::net::each_addr<> core::result::Result<>,std::sys_common::net::UdpSocket> [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\net\mod.rs:113]
# 8 std::net::udp::UdpSocket::bind<>                                           [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\net\udp.rs:94]
# 9 mio::net::udp::UdpSocket::bind                                             [C:\Users\dennis\.cargo\git\checkouts\mio-76192d34136afce1\7358ab7\src\net\udp.rs:122]
#10 mio_mao::main                                                              [G:\Dev\test-mio-mao\src\main.rs:15]
#11 std::rt::lang_start::{{closure}}<>                                         [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
#12 std::rt::lang_start_internal                                               [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\rt.rs:48]
#13 std::rt::lang_start<>                                                      [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
#14 main       
Note: @0:00:01.083 in thread 31520
Note: instruction: cmovo  %r13 -> %rax

Error #3: UNINITIALIZED READ: reading 0x0000001dbcefe8c0-0x0000001dbcefe8c4 4 byte(s) within 0x0000001dbcefe8c0-0x0000001dbcefe8c8
# 0 replace_memset                                                             [d:\drmemory_package\drmemory\replace.c:200]
# 1 WS2_32.dll!WSALookupServiceNextW                                          +0x3088   (0x00007ffdca6eee29 <WS2_32.dll+0xee29>)
# 2 WS2_32.dll!WSALookupServiceNextW                                          +0x3c3d   (0x00007ffdca6ef9de <WS2_32.dll+0xf9de>)
# 3 WS2_32.dll!WSAStartup                                                     +0x33c    (0x00007ffdca6efd4d <WS2_32.dll+0xfd4d>)
# 4 std::sync::once::{{impl}}::call_once::{{closure}}<>                        [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\sync\once.rs:220]
# 5 std::sync::once::Once::call_inner                                          [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\sync\once.rs:387]
# 6 std::sys_common::net::UdpSocket::bind                                      [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\sys_common\net.rs:468]
# 7 core::ops::function::FnMut::call_mut<> core::result::Result<>,(core::result::Result<>)> [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libcore\ops\function.rs:148]
# 8 std::net::each_addr<> core::result::Result<>,std::sys_common::net::UdpSocket> [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\net\mod.rs:113]
# 9 std::net::udp::UdpSocket::bind<>                                           [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\net\udp.rs:94]
#10 mio::net::udp::UdpSocket::bind                                             [C:\Users\dennis\.cargo\git\checkouts\mio-76192d34136afce1\7358ab7\src\net\udp.rs:122]
#11 mio_mao::main                                                              [G:\Dev\test-mio-mao\src\main.rs:15]
#12 std::rt::lang_start::{{closure}}<>                                         [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
#13 std::rt::lang_start_internal                                               [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\rt.rs:48]
#14 std::rt::lang_start<>                                                      [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
#15 main       
Note: @0:00:01.105 in thread 31520
Note: instruction: cmp    0x30(%rsp) $0x0000000000000003

Error #4: UNINITIALIZED READ: reading 0x0000001dbcefea4c-0x0000001dbcefea50 4 byte(s) within 0x0000001dbcefe9d0-0x0000001dbcefea54
# 0 system call NtDeviceIoControlFile SOCKET_CONTEXT pre-Padding
# 1 MSWSOCK.dll!NSPStartup                                                    +0x595c   (0x00007ffdc8617afd <MSWSOCK.dll+0x7afd>)
# 2 MSWSOCK.dll!NSPStartup                                                    +0x3d4b   (0x00007ffdc8615eec <MSWSOCK.dll+0x5eec>)
# 3 WS2_32.dll!WSASocketW                                                     +0x115    (0x00007ffdca6e9e16 <WS2_32.dll+0x9e16>)
# 4 std::sys::windows::net::Socket::new                                        [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\sys\windows\net.rs:99]
# 5 std::sys_common::net::UdpSocket::bind                                      [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\sys_common\net.rs:465]
# 6 core::ops::function::FnMut::call_mut<> core::result::Result<>,(core::result::Result<>)> [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libcore\ops\function.rs:148]
# 7 std::net::each_addr<> core::result::Result<>,std::sys_common::net::UdpSocket> [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\net\mod.rs:113]
# 8 std::net::udp::UdpSocket::bind<>                                           [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\net\udp.rs:94]
# 9 mio::net::udp::UdpSocket::bind                                             [C:\Users\dennis\.cargo\git\checkouts\mio-76192d34136afce1\7358ab7\src\net\udp.rs:122]
#10 mio_mao::main                                                              [G:\Dev\test-mio-mao\src\main.rs:15]
#11 std::rt::lang_start::{{closure}}<>                                         [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
#12 std::rt::lang_start_internal                                               [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\rt.rs:48]
#13 std::rt::lang_start<>                                                      [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
#14 main       
Note: @0:00:01.435 in thread 31520

Error #5: UNINITIALIZED READ: reading 0x0000001dbcefeabc-0x0000001dbcefeac0 4 byte(s) within 0x0000001dbcefea40-0x0000001dbcefeac4
# 0 system call NtDeviceIoControlFile SOCKET_CONTEXT pre-Padding
# 1 MSWSOCK.dll!NSPStartup                                                    +0x595c   (0x00007ffdc8617afd <MSWSOCK.dll+0x7afd>)
# 2 MSWSOCK.dll!NSPStartup                                                    +0x5d45   (0x00007ffdc8617ee6 <MSWSOCK.dll+0x7ee6>)
# 3 WS2_32.dll!bind                                                           +0x6e     (0x00007ffdca6f1ddf <WS2_32.dll+0x11ddf>)
# 4 std::sys_common::net::UdpSocket::bind                                      [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\sys_common\net.rs:467]
# 5 core::ops::function::FnMut::call_mut<> core::result::Result<>,(core::result::Result<>)> [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libcore\ops\function.rs:148]
# 6 std::net::each_addr<> core::result::Result<>,std::sys_common::net::UdpSocket> [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\net\mod.rs:113]
# 7 std::net::udp::UdpSocket::bind<>                                           [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\net\udp.rs:94]
# 8 mio::net::udp::UdpSocket::bind                                             [C:\Users\dennis\.cargo\git\checkouts\mio-76192d34136afce1\7358ab7\src\net\udp.rs:122]
# 9 mio_mao::main                                                              [G:\Dev\test-mio-mao\src\main.rs:15]
#10 std::rt::lang_start::{{closure}}<>                                         [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
#11 std::rt::lang_start_internal                                               [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\rt.rs:48]
#12 std::rt::lang_start<>                                                      [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
#13 main       
Note: @0:00:01.463 in thread 31520

Error #6: UNINITIALIZED READ: reading 0x0000001dbcefef3c-0x0000001dbcefef40 4 byte(s) within 0x0000001dbcefef38-0x0000001dbcefef48
# 0 system call NtDeviceIoControlFile AFD_SEND_INFO_UDP.BufferArray
# 1 MSWSOCK.dll!Tcpip6_WSHGetWildcardSockaddr                                 +0x417    (0x00007ffdc861f9f8 <MSWSOCK.dll+0xf9f8>)
# 2 WS2_32.dll!sendto                                                         +0xe9     (0x00007ffdca6f16ca <WS2_32.dll+0x116ca>)
# 3 std::sys_common::net::UdpSocket::send_to                                   [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\sys_common\net.rs:499]
# 4 std::net::udp::UdpSocket::send_to<>                                        [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\net\udp.rs:177]
# 5 mio::sys::windows::udp::UdpSocket::send_to                                 [C:\Users\dennis\.cargo\git\checkouts\mio-76192d34136afce1\7358ab7\src\sys\windows\udp.rs:78]
# 6 mio::net::udp::UdpSocket::send_to                                          [C:\Users\dennis\.cargo\git\checkouts\mio-76192d34136afce1\7358ab7\src\net\udp.rs:224]
# 7 mio_mao::main                                                              [G:\Dev\test-mio-mao\src\main.rs:20]
# 8 std::rt::lang_start::{{closure}}<>                                         [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
# 9 std::rt::lang_start_internal                                               [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\rt.rs:48]
#10 std::rt::lang_start<>                                                      [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
#11 main       
Note: @0:00:01.721 in thread 31520

Error #7: UNINITIALIZED READ: reading 0x0000001dbcefee14-0x0000001dbcefee18 4 byte(s) within 0x0000001dbcefee14-0x0000001dbcefee18
# 0 system call NtDeviceIoControlFile AFD_SEND_INFO_UDP.SizeOfRemoteAddress
# 1 MSWSOCK.dll!Tcpip6_WSHGetWildcardSockaddr                                 +0x417    (0x00007ffdc861f9f8 <MSWSOCK.dll+0xf9f8>)
# 2 WS2_32.dll!sendto                                                         +0xe9     (0x00007ffdca6f16ca <WS2_32.dll+0x116ca>)
# 3 std::sys_common::net::UdpSocket::send_to                                   [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\sys_common\net.rs:499]
# 4 std::net::udp::UdpSocket::send_to<>                                        [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\net\udp.rs:177]
# 5 mio::sys::windows::udp::UdpSocket::send_to                                 [C:\Users\dennis\.cargo\git\checkouts\mio-76192d34136afce1\7358ab7\src\sys\windows\udp.rs:78]
# 6 mio::net::udp::UdpSocket::send_to                                          [C:\Users\dennis\.cargo\git\checkouts\mio-76192d34136afce1\7358ab7\src\net\udp.rs:224]
# 7 mio_mao::main                                                              [G:\Dev\test-mio-mao\src\main.rs:20]
# 8 std::rt::lang_start::{{closure}}<>                                         [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
# 9 std::rt::lang_start_internal                                               [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\rt.rs:48]
#10 std::rt::lang_start<>                                                      [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
#11 main       
Note: @0:00:01.735 in thread 31520

Error #8: UNINITIALIZED READ: reading 0x0000001dbcefee18-0x0000001dbcefee20 8 byte(s) within 0x0000001dbcefee18-0x0000001dbcefee20
# 0 system call NtDeviceIoControlFile AFD_SEND_INFO_UDP.RemoteAddress
# 1 MSWSOCK.dll!Tcpip6_WSHGetWildcardSockaddr                                 +0x417    (0x00007ffdc861f9f8 <MSWSOCK.dll+0xf9f8>)
# 2 WS2_32.dll!sendto                                                         +0xe9     (0x00007ffdca6f16ca <WS2_32.dll+0x116ca>)
# 3 std::sys_common::net::UdpSocket::send_to                                   [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\sys_common\net.rs:499]
# 4 std::net::udp::UdpSocket::send_to<>                                        [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\net\udp.rs:177]
# 5 mio::sys::windows::udp::UdpSocket::send_to                                 [C:\Users\dennis\.cargo\git\checkouts\mio-76192d34136afce1\7358ab7\src\sys\windows\udp.rs:78]
# 6 mio::net::udp::UdpSocket::send_to                                          [C:\Users\dennis\.cargo\git\checkouts\mio-76192d34136afce1\7358ab7\src\net\udp.rs:224]
# 7 mio_mao::main                                                              [G:\Dev\test-mio-mao\src\main.rs:20]
# 8 std::rt::lang_start::{{closure}}<>                                         [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
# 9 std::rt::lang_start_internal                                               [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\rt.rs:48]
#10 std::rt::lang_start<>                                                      [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
#11 main       
Note: @0:00:01.738 in thread 31520

Error #9: LEAK 40 direct bytes 0x00000198a8ea07c0-0x00000198a8ea07e8 + 8 indirect bytes
# 0 replace_RtlAllocateHeap                    [d:\drmemory_package\common\alloc_replace.c:3771]
# 1 ntdll.dll!RtlpAddVectoredHandler
# 2 std::rt::lang_start_internal               [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\/src\libstd\rt.rs:34]
# 3 std::rt::lang_start<>                      [/rustc/a53f9df32fbb0b5f4382caaad8f1a46f36ea887c\src\libstd\rt.rs:64]
# 4 main       

===========================================================================
FINAL SUMMARY:

DUPLICATE ERROR COUNTS:

SUPPRESSIONS USED:

ERRORS FOUND:
      0 unique,     0 total unaddressable access(es)
      8 unique,     8 total uninitialized access(es)
      0 unique,     0 total invalid heap argument(s)
      0 unique,     0 total GDI usage error(s)
      0 unique,     0 total handle leak(s)
      0 unique,     0 total warning(s)
      1 unique,     1 total,     48 byte(s) of leak(s)
      0 unique,     0 total,      0 byte(s) of possible leak(s)
ERRORS IGNORED:
     24 potential error(s) (suspected false positives)
         (details: C:\Users\dennis\AppData\Roaming\Dr. Memory\DrMemory-mio-mao.exe.31512.000\potential_errors.txt)
     22 unique,    22 total,   3585 byte(s) of still-reachable allocation(s)
         (re-run with "-show_reachable" for details)
Details: C:\Users\dennis\AppData\Roaming\Dr. Memory\DrMemory-mio-mao.exe.31512.000\results.txt
```

Reply:
The "apccontext" pointer as such is treated as an opaque value by the kernel.

(The actual value of the apccontext here is a pointer to an ntstatus block embedded in an overlapped. The kernel does take ownership of the ntstatus until the operation completes)

Reply:
@piscisaureus right, but what I mean is that **we** are passing a pointer in... unless we never deref it again (which would be surprising), we need to make sure the lifetimes are correct. I would assume we get the pointer back when the afd event completes?

Reply:
There are only two operations that use the pointer. update and cancel in the socket state. The socket state is being included by Mutex. So the operation can only be happening one time. Update can be used at start of the poll. Cancel can be used while the polling. And the pointer isn't being freed if the operation is pending.
Of course my code is just copying the implementation of wepoll source code. So there's maybe a better refactor structure we can do.